File: PracticasProgramacion/Practica_12/cpp/Matriz.cpp
#include "Matriz.h"
#include <iostream>
#include <stdexcept>

Matriz::Matriz(int filas, int columnas) : filas_(filas), columnas_(columnas) {
    datos_ = new double*[filas_];
    for (int i = 0; i < filas_; ++i)
        datos_[i] = new double[columnas_]();
}

Matriz::Matriz(const Matriz& otra) : filas_(otra.filas_), columnas_(otra.columnas_) {
    datos_ = new double*[filas_];
    for (int i = 0; i < filas_; ++i) {
        datos_[i] = new double[columnas_];
        for (int j = 0; j < columnas_; ++j)
            datos_[i][j] = otra.datos_[i][j];
    }
}

Matriz& Matriz::operator=(const Matriz& otra) {
    if (this != &otra) {
        liberar();
        filas_ = otra.filas_;
        columnas_ = otra.columnas_;
        datos_ = new double*[filas_];
        for (int i = 0; i < filas_; ++i) {
            datos_[i] = new double[columnas_];
            for (int j = 0; j < columnas_; ++j)
                datos_[i][j] = otra.datos_[i][j];
        }
    }
    return *this;
}

Matriz::~Matriz() { liberar(); }

void Matriz::liberar() {
    if (datos_) {
        for (int i = 0; i < filas_; ++i)
            delete[] datos_[i];
        delete[] datos_;
        datos_ = nullptr;
    }
}

void Matriz::llenar(double valor) {
    for (int i = 0; i < filas_; ++i)
        for (int j = 0; j < columnas_; ++j)
            datos_[i][j] = valor;
}

void Matriz::set(int f, int c, double v) { datos_[f][c] = v; }
double Matriz::get(int f, int c) const { return datos_[f][c]; }

Matriz Matriz::multiplicarPorConstante(double k) const {
    Matriz R(filas_, columnas_);
    for (int i = 0; i < filas_; ++i)
        for (int j = 0; j < columnas_; ++j)
            R.datos_[i][j] = datos_[i][j] * k;
    return R;
}

Matriz Matriz::multiplicar(const Matriz& B) const {
    if (columnas_ != B.filas_)
        throw std::runtime_error("Dimensiones incompatibles");
    Matriz R(filas_, B.columnas_);
    for (int i = 0; i < filas_; ++i)
        for (int j = 0; j < B.columnas_; ++j) {
            double s = 0;
            for (int k = 0; k < columnas_; ++k)
                s += datos_[i][k] * B.datos_[k][j];
            R.datos_[i][j] = s;
        }
    return R;
}

void Matriz::imprimir() const {
    for (int i = 0; i < filas_; ++i) {
        for (int j = 0; j < columnas_; ++j)
            std::cout << datos_[i][j] << " ";
        std::cout << "\n";
    }
}

File: PracticasProgramacion/Practica_12/python/matriz.py
"""
Matriz: producto por escalar y producto matricial.
datos_ como lista de listas; producto A*B exige columnas_A == filas_B.
"""


class Matriz:
    def __init__(self, filas: int, columnas: int) -> None:
        self._filas = filas
        self._columnas = columnas
        self._datos = [[0.0] * columnas for _ in range(filas)]

    def llenar(self, valor: float) -> None:
        for i in range(self._filas):
            for j in range(self._columnas):
                self._datos[i][j] = valor

    def set_val(self, f: int, c: int, v: float) -> None:
        self._datos[f][c] = v

    def get_val(self, f: int, c: int) -> float:
        return self._datos[f][c]

    def multiplicar_por_constante(self, k: float) -> "Matriz":
        r = Matriz(self._filas, self._columnas)
        for i in range(self._filas):
            for j in range(self._columnas):
                r._datos[i][j] = self._datos[i][j] * k
        return r

    def multiplicar(self, b: "Matriz") -> "Matriz":
        if self._columnas != b._filas:
            raise ValueError("Dimensiones incompatibles")
        r = Matriz(self._filas, b._columnas)
        for i in range(self._filas):
            for j in range(b._columnas):
                s = 0.0
                for k in range(self._columnas):
                    s += self._datos[i][k] * b._datos[k][j]
                r._datos[i][j] = s
        return r

    def imprimir(self) -> None:
        for i in range(self._filas):
            print(" ".join(str(self._datos[i][j]) for j in range(self._columnas)))
